src/tfs/tlog.c

```diff
@@ -1,180 +1,418 @@
 #include <tfs_internal.h>
 
 //#define TLOG_DEBUG
+//#define TLOG_DEBUG_DUMP
 #ifdef TLOG_DEBUG
 #define tlog_debug(x, ...) do {rprintf("TLOG: " x, ##__VA_ARGS__);} while(0)
 #else
 #define tlog_debug(x, ...)
 #endif
 
+#define TFS_MAGIC_BYTES 6
+static const char *tfs_magic = "NVMTFS";
+
 #define END_OF_LOG 1
 #define TUPLE_AVAILABLE 2
-#define END_OF_SEGMENT 3
+#define TUPLE_EXTENDED 3
+#define END_OF_SEGMENT 4
+#define LOG_EXTENSION_LINK 5
+
+#define COMPLETION_QUEUE_SIZE 10
+
+#define MAX_VARINT_SIZE 10 /* to encode 64 significant bits */
+
+#define TFS_EXTENSION_HEADER_BYTES (TFS_MAGIC_BYTES + 2 * MAX_VARINT_SIZE)
+#define TFS_EXTENSION_LINK_BYTES (1 + 2 * MAX_VARINT_SIZE)
+#define TFS_LOG_RESERVED_BYTES (TFS_EXTENSION_HEADER_BYTES + TFS_EXTENSION_LINK_BYTES)
 
 typedef struct log {
     filesystem fs;
-    u64 remainder;
-    buffer staging;
     vector completions;
     table dictionary;
-    u64 offset;
+
+    /* sector offset, length and staging buffer of current extension */
+    range sectors;
+    buffer staging;
+    buffer tuple_staging;
+    u64 tuple_bytes_remain;
+    boolean extension_open;
+
     int dirty;              /* cas boolean */
     heap h;
 } *log;
 
-closure_function(1, 1, void, log_write_completion,
-                 vector, v,
+closure_function(3, 1, void, log_write_completion,
+                 buffer, b,
+                 vector, completions,
+                 boolean, release,
                  status, s)
 {
     // reclaim the buffer now and the vector...make it a whole thing
-    status_handler i;
-    int len = vector_length(bound(v));
-    for (int count = 0; count < len; count++) {
-        i = vector_delete(bound(v), 0);
-        apply(i, s);
-    }
+    vector v = bound(completions);
+    status_handler sh;
+
+    vector_foreach(v, sh)
+        apply(sh, s);
+
+    deallocate_vector(v);
+
+    if (bound(release))
+        deallocate_buffer(bound(b));
     closure_finish();
 }
 
 // xxx  currently we cant take writes during the flush
 
-/* XXX it's not right to just stick SECTOR_{SIZE,OFFSET} everywhere...
-   and add block_log2 to fs */
-range log_block_range(log tl, u64 length)
+/* Avoid references to log, which may be in transition to a new extension. */
```

Comment:
Sure, that would make sense...and it could be refcounted. But it seemed to me like the flush is the singular terminal point for such a segment...so it would be creating a type just to avoid some messy argument lists, I think. Of course, that's if we continue to accept having total amnesia of parsed extensions.